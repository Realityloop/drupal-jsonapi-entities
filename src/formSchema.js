import schemaBase from './schemaBase'
import entityFormDisplay from './resources/entityFormDisplay'
import fieldConfig from './resources/fieldConfig'
import fieldStorageConfig from './resources/fieldStorageConfig'

class formSchema extends schemaBase {
  constructor(entityType, bundle, mode, context) {
    super(entityType, bundle, mode, context)

    this.resources = [
      new entityFormDisplay(this),
      new fieldConfig(this),
      new fieldStorageConfig(this)
    ]
  }

  validate() {
    const entityFormDisplay = this.data[this.resources[0].requestId]

    if (entityFormDisplay.length === 0) {
      throw new Error(`Entity type invalid: ${this.entityType} ${this.bundle}`)
    }
    return true
  }

  processFields() {
    const fields = {}

    for (const resource of this.resources) {
      for (const field in resource.fields) {
        if (typeof fields[field] === 'undefined' && resource instanceof entityFormDisplay) {
          fields[field] = resource.fields[field]
        }

        else if (typeof fields[field] !== 'undefined') {
          // Remove settings, they're merged seperately.
          const settings = resource.fields[field].settings
          delete resource.fields[field].settings

          fields[field] = Object.assign(fields[field], resource.fields[field])
          fields[field].settings = Object.assign(fields[field].settings, settings)
        }
      }
    }

    return fields
  }

  processGroups() {
    return this.resources[0].groups
  }
}

export default formSchema
