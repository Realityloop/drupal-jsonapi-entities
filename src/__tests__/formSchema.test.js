import formSchema from '../formSchema'

import mockEntityFormDisplay from '../resources/__fixtures__/entity_form_display'
import mockFieldConfig from '../resources/__fixtures__/field_config'
import mockFieldStorageConfig from '../resources/__fixtures__/field_storage_config'

const entityType = 'node'
const bundle = 'recipe'
const mode = 'default'

const mockAxios = {
  post: () => ({
    data: {
      'node--recipe--default--entity_form_display': { body: JSON.stringify(mockEntityFormDisplay) },
      'node--recipe--field_config': { body: JSON.stringify(mockFieldConfig) },
      'node--field_storage_config': { body: JSON.stringify(mockFieldStorageConfig) }
    }
  })
}

test('formSchema', async () => {
  const schema = new formSchema({ entityType, bundle, mode, axios: mockAxios })
  await schema.build()

  expect(schema.fields.items).toMatchSnapshot()
  expect(schema.keys).toMatchSnapshot()
  expect(schema.fields.get()).toMatchSnapshot()
  expect(schema.fields.get(false)).toMatchSnapshot()
})
