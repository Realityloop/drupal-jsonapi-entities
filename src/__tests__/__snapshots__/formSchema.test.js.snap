// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formSchema 1`] = `
Object {
  "created": Object {
    "id": "created",
    "property": true,
    "settings": Object {},
    "type": "datetime_timestamp",
    "weight": 13,
  },
  "field_author": Object {
    "cardinality": 1,
    "description": "The author of this recipe. Start typing their username to search.",
    "id": "field_author",
    "label": "Author",
    "property": false,
    "required": true,
    "settings": Object {
      "handler": "default:user",
      "handler_settings": Object {
        "auto_create": false,
        "filter": Object {
          "type": "_none",
        },
        "include_anonymous": true,
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": null,
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "user",
    },
    "type": "entity_reference_autocomplete",
    "weight": 1,
  },
  "field_cooking_time": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 3,
  },
  "field_difficulty": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "property": false,
    "required": true,
    "settings": Object {
      "allowed_values": Object {
        "easy": "Easy",
        "hard": "Hard",
        "medium": "Medium",
      },
      "allowed_values_function": "",
    },
    "type": "options_select",
    "weight": 5,
  },
  "field_image": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_image",
    "label": "Image",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "media_types": Array [],
      "target_type": "media",
    },
    "type": "media_library_widget",
    "weight": 8,
  },
  "field_ingredients": Object {
    "cardinality": -1,
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "property": false,
    "required": false,
    "settings": Object {
      "case_sensitive": false,
      "is_ascii": false,
      "max_length": 255,
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 10,
  },
  "field_number_of_servings": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": "",
      "unsigned": false,
    },
    "type": "number",
    "weight": 4,
  },
  "field_preparation_time": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 2,
  },
  "field_recipe_category": Object {
    "cardinality": -1,
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "target_type": "taxonomy_term",
    },
    "type": "options_select",
    "weight": 6,
  },
  "field_recipe_instruction": Object {
    "cardinality": 1,
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 11,
  },
  "field_summary": Object {
    "cardinality": 1,
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 9,
  },
  "field_tags": Object {
    "cardinality": -1,
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "taxonomy_term",
    },
    "type": "entity_reference_autocomplete_tags",
    "weight": 7,
  },
  "langcode": Object {
    "id": "langcode",
    "property": true,
    "settings": Object {
      "include_locked": true,
    },
    "type": "language_select",
    "weight": 2,
  },
  "path": Object {
    "id": "path",
    "property": true,
    "settings": Object {},
    "type": "path",
    "weight": 16,
  },
  "promote": Object {
    "id": "promote",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 14,
  },
  "status": Object {
    "id": "status",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 17,
  },
  "sticky": Object {
    "id": "sticky",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 15,
  },
  "title": Object {
    "id": "title",
    "property": true,
    "settings": Object {
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 0,
  },
  "uid": Object {
    "id": "uid",
    "property": true,
    "settings": Object {
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
    },
    "type": "entity_reference_autocomplete",
    "weight": 12,
  },
}
`;

exports[`formSchema 2`] = `
Array [
  "created",
  "field_author",
  "field_cooking_time",
  "field_difficulty",
  "field_image",
  "field_ingredients",
  "field_number_of_servings",
  "field_preparation_time",
  "field_recipe_category",
  "field_recipe_instruction",
  "field_summary",
  "field_tags",
  "langcode",
  "path",
  "promote",
  "status",
  "sticky",
  "title",
  "uid",
]
`;

exports[`formSchema 3`] = `
Array [
  Object {
    "id": "title",
    "property": true,
    "settings": Object {
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 0,
  },
  Object {
    "cardinality": 1,
    "description": "The author of this recipe. Start typing their username to search.",
    "id": "field_author",
    "label": "Author",
    "property": false,
    "required": true,
    "settings": Object {
      "handler": "default:user",
      "handler_settings": Object {
        "auto_create": false,
        "filter": Object {
          "type": "_none",
        },
        "include_anonymous": true,
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": null,
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "user",
    },
    "type": "entity_reference_autocomplete",
    "weight": 1,
  },
  Object {
    "id": "langcode",
    "property": true,
    "settings": Object {
      "include_locked": true,
    },
    "type": "language_select",
    "weight": 2,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 2,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 3,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": "",
      "unsigned": false,
    },
    "type": "number",
    "weight": 4,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "property": false,
    "required": true,
    "settings": Object {
      "allowed_values": Object {
        "easy": "Easy",
        "hard": "Hard",
        "medium": "Medium",
      },
      "allowed_values_function": "",
    },
    "type": "options_select",
    "weight": 5,
  },
  Object {
    "cardinality": -1,
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "target_type": "taxonomy_term",
    },
    "type": "options_select",
    "weight": 6,
  },
  Object {
    "cardinality": -1,
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "taxonomy_term",
    },
    "type": "entity_reference_autocomplete_tags",
    "weight": 7,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_image",
    "label": "Image",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "media_types": Array [],
      "target_type": "media",
    },
    "type": "media_library_widget",
    "weight": 8,
  },
  Object {
    "cardinality": 1,
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 9,
  },
  Object {
    "cardinality": -1,
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "property": false,
    "required": false,
    "settings": Object {
      "case_sensitive": false,
      "is_ascii": false,
      "max_length": 255,
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 10,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 11,
  },
  Object {
    "id": "uid",
    "property": true,
    "settings": Object {
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
    },
    "type": "entity_reference_autocomplete",
    "weight": 12,
  },
  Object {
    "id": "created",
    "property": true,
    "settings": Object {},
    "type": "datetime_timestamp",
    "weight": 13,
  },
  Object {
    "id": "promote",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 14,
  },
  Object {
    "id": "sticky",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 15,
  },
  Object {
    "id": "path",
    "property": true,
    "settings": Object {},
    "type": "path",
    "weight": 16,
  },
  Object {
    "id": "status",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 17,
  },
]
`;

exports[`formSchema 4`] = `
Array [
  Object {
    "id": "created",
    "property": true,
    "settings": Object {},
    "type": "datetime_timestamp",
    "weight": 13,
  },
  Object {
    "cardinality": 1,
    "description": "The author of this recipe. Start typing their username to search.",
    "id": "field_author",
    "label": "Author",
    "property": false,
    "required": true,
    "settings": Object {
      "handler": "default:user",
      "handler_settings": Object {
        "auto_create": false,
        "filter": Object {
          "type": "_none",
        },
        "include_anonymous": true,
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": null,
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "user",
    },
    "type": "entity_reference_autocomplete",
    "weight": 1,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_cooking_time",
    "label": "Cooking time",
    "property": false,
    "required": false,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 3,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_difficulty",
    "label": "Difficulty",
    "property": false,
    "required": true,
    "settings": Object {
      "allowed_values": Object {
        "easy": "Easy",
        "hard": "Hard",
        "medium": "Medium",
      },
      "allowed_values_function": "",
    },
    "type": "options_select",
    "weight": 5,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_image",
    "label": "Image",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:media",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "field": "_none",
        },
        "target_bundles": Object {
          "image": "image",
        },
      },
      "media_types": Array [],
      "target_type": "media",
    },
    "type": "media_library_widget",
    "weight": 8,
  },
  Object {
    "cardinality": -1,
    "description": "List the ingredients required for this recipe, one per item.",
    "id": "field_ingredients",
    "label": "Ingredients",
    "property": false,
    "required": false,
    "settings": Object {
      "case_sensitive": false,
      "is_ascii": false,
      "max_length": 255,
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 10,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_number_of_servings",
    "label": "Number of servings",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": "",
      "unsigned": false,
    },
    "type": "number",
    "weight": 4,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_preparation_time",
    "label": "Preparation time",
    "property": false,
    "required": true,
    "settings": Object {
      "max": null,
      "min": 0,
      "placeholder": "",
      "prefix": "",
      "size": "normal",
      "suffix": " minutes",
      "unsigned": false,
    },
    "type": "number",
    "weight": 2,
  },
  Object {
    "cardinality": -1,
    "description": "",
    "id": "field_recipe_category",
    "label": "Recipe category",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "recipe_category": "recipe_category",
        },
      },
      "target_type": "taxonomy_term",
    },
    "type": "options_select",
    "weight": 6,
  },
  Object {
    "cardinality": 1,
    "description": "",
    "id": "field_recipe_instruction",
    "label": "Recipe instruction",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 11,
  },
  Object {
    "cardinality": 1,
    "description": "Provide a short overview of this recipe.",
    "id": "field_summary",
    "label": "Summary",
    "property": false,
    "required": true,
    "settings": Object {
      "placeholder": "",
      "rows": 5,
    },
    "type": "text_textarea",
    "weight": 9,
  },
  Object {
    "cardinality": -1,
    "description": "",
    "id": "field_tags",
    "label": "Tags",
    "property": false,
    "required": false,
    "settings": Object {
      "handler": "default:taxonomy_term",
      "handler_settings": Object {
        "auto_create": false,
        "auto_create_bundle": "",
        "sort": Object {
          "direction": "asc",
          "field": "name",
        },
        "target_bundles": Object {
          "tags": "tags",
        },
      },
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
      "target_type": "taxonomy_term",
    },
    "type": "entity_reference_autocomplete_tags",
    "weight": 7,
  },
  Object {
    "id": "langcode",
    "property": true,
    "settings": Object {
      "include_locked": true,
    },
    "type": "language_select",
    "weight": 2,
  },
  Object {
    "id": "path",
    "property": true,
    "settings": Object {},
    "type": "path",
    "weight": 16,
  },
  Object {
    "id": "promote",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 14,
  },
  Object {
    "id": "status",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 17,
  },
  Object {
    "id": "sticky",
    "property": true,
    "settings": Object {
      "display_label": true,
    },
    "type": "boolean_checkbox",
    "weight": 15,
  },
  Object {
    "id": "title",
    "property": true,
    "settings": Object {
      "placeholder": "",
      "size": 60,
    },
    "type": "string_textfield",
    "weight": 0,
  },
  Object {
    "id": "uid",
    "property": true,
    "settings": Object {
      "match_operator": "CONTAINS",
      "placeholder": "",
      "size": 60,
    },
    "type": "entity_reference_autocomplete",
    "weight": 12,
  },
]
`;
